src/beast/beast/http/basic_parser.h

```diff
@@ -113,153 +133,405 @@ class basic_parser
+    {
+        template<class T, class R =
+            decltype(std::declval<T>().on_body(
+                std::declval<void const*>(), std::declval<std::size_t>(),
+                    std::declval<error_code&>()), std::true_type{})>
+        static R check(int);
+        template <class>
+        static std::false_type check(...);
+        using type = decltype(check<C>(0));
+    public:
+        static bool const value = type::value;
+    };
+    template<class C>
+    using has_on_body =
+        std::bool_constant<has_on_body_t<C>::value>;
 
-        @param data A memory block containing the next decoded
-                    chunk of the content body.
-        @param bytes The number of bytes pointed to by data.
-    */
-    virtual
     void
-    on_body (void const* data, std::size_t bytes) = 0;
+    call_on_body(void const* data, std::size_t bytes,
+        error_code& ec, std::true_type)
+    {
+        impl().on_body(data, bytes, ec);
+    }
 
-    /** Called once when the message is complete. */
-    virtual
     void
-    on_complete() = 0;
+    call_on_body(void const*, std::size_t,
+        error_code&, std::false_type)
+    {
+    }
 
-private:
-    void check_header();
-
-    int do_message_start ();
-    int do_url (char const* in, std::size_t bytes);
-    int do_status (char const* in, std::size_t bytes);
-    int do_header_field (char const* in, std::size_t bytes);
-    int do_header_value (char const* in, std::size_t bytes);
-    int do_headers_complete ();
-    int do_body (char const* in, std::size_t bytes);
-    int do_message_complete ();
-    int do_chunk_header();
-    int do_chunk_complete();
-
-    static int cb_message_start (joyent::http_parser*);
-    static int cb_url (joyent::http_parser*, char const*, std::size_t);
-    static int cb_status (joyent::http_parser*, char const*, std::size_t);
-    static int cb_header_field (joyent::http_parser*, char const*, std::size_t);
-    static int cb_header_value (joyent::http_parser*, char const*, std::size_t);
-    static int cb_headers_complete (joyent::http_parser*);
-    static int cb_body (joyent::http_parser*, char const*, std::size_t);
-    static int cb_message_complete (joyent::http_parser*);
-    static int cb_chunk_header (joyent::http_parser*);
-    static int cb_chunk_complete (joyent::http_parser*);
+    template<class C>
+    class has_on_complete_t
+    {
+        template<class T, class R =
+            decltype(std::declval<T>().on_complete(), std::true_type{})>
+        static R check(int);
+        template <class>
+        static std::false_type check(...);
+        using type = decltype(check<C>(0));
+    public:
+        static bool const value = type::value;
+    };
+    template<class C>
+    using has_on_complete =
+        std::bool_constant<has_on_complete_t<C>::value>;
+
+    void
+    call_on_complete(std::true_type)
+    {
+        impl().on_complete();
+    }
+
+    void
+    call_on_complete(std::false_type)
+    {
+    }
+
```

Comment:
General style comment on all of these `call_on` functions.  You can make them slightly less verbose at the call site by adding this overload:
````
void call_on_complete() {call_on_complete(has_on_complete<Derived>{});}
````
So your use code now looks like:
````
t.call_on_complete();
````
Not a huge deal either way since these are all private functions.  This just helps encapsulate the details closer together.